File: 2023/01. Trebuchet/c++/two/src/main.cpp
#include <algorithm>
#include <array>
#include <cstddef>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <ranges>
#include <string>
#include <string_view>

static constexpr std::array<std::string_view, 9> DIGIT_STRINGS{
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine"
};

int main(const int argc, const char** const argv)
{
    if (argc != 2)
    {
        std::cerr << "Usage: " << argv[0] << " <INPUT_FILE>\n";
        return 1;
    }

    std::ifstream infile{argv[1]};
    if (!infile.is_open())
    {
        std::cerr << "ERROR! Failed to open " << std::quoted(argv[1]) << '\n';
        return 1;
    }

    int total{};

    for (std::string line; std::getline(infile, line);)
    {
        // Find first actual digit
        auto firstIt{std::ranges::find_if(line, [](const char c) -> bool
        {
            return std::isdigit(c);
        })};

        int first;

        // Search for every digit string and replace firstIt and set first if it's found closer to the begining
        for (std::size_t i{}; i < DIGIT_STRINGS.size(); ++i)
        {
            const auto digitString{DIGIT_STRINGS[i]};

            const std::size_t pos{line.find(digitString)};
            if (pos == std::string::npos)
                continue;

            const auto it{line.begin() + static_cast<std::ptrdiff_t>(pos)};

            if (it < firstIt)
            {
                firstIt = it;
                first = static_cast<int>(i) + 1;
            }
        }

        // Find last digit
        auto lastIt{std::ranges::find_if(std::views::reverse(line), [](const char c) -> bool
        {
            return std::isdigit(c);
        })};
        
        int last;

        // Search for every digit string and replace lastIt and set last if it's found closer to the begining
        for (std::size_t i{}; i < DIGIT_STRINGS.size(); ++i)
        {
            const auto digitString{DIGIT_STRINGS[i]};

            const std::size_t pos{line.rfind(digitString)};
            if (pos == std::string::npos)
                continue;

            // Extra math needed to convert from index to reverse iterator =
            const auto it{line.rbegin() + static_cast<std::ptrdiff_t>(line.size()) - 1 - static_cast<std::ptrdiff_t>(pos)};

            if (it < lastIt)
            {
                lastIt = it;
                last = static_cast<int>(i) + 1;
            }
        }

        // If the it's don't point to digits, they point to digit strings and the int representations has already been set
        if (std::isdigit(*firstIt))
            first = *firstIt - '0';
        if (std::isdigit(*lastIt))
            last = *lastIt - '0';

        // Assemble into full number and add to running total
        const int assembled{(first * 10) + last};
        total += assembled;
    }

    std::cout << "Result: " << total << '\n'; 

    return 0;
}

File: 2023/01. Trebuchet/rust/two/src/main.rs
use std::env;
use std::fs;

const DIGIT_STRINGS: [&str; 9] = [
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine"
];

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {}", args[0]);
        return;
    }

    let mut total = 0;

    // for each line in file
    let contents = fs::read_to_string(&args[1]).expect("Could not read input file");
    for line in contents.lines() {
        // u32::MAX represents not yet found
        let mut first_digit = u32::MAX;
        let mut last_digit = u32::MAX;

        // Look for first numeric digit
        let mut first_digit_index = line
            .find(|c: char| c.is_ascii_digit())
            .unwrap_or(usize::MAX);

        // If numeric digit was found convert from char to numeric
        if first_digit_index != usize::MAX {
            first_digit = line
                .chars()
                .nth(first_digit_index)
                .unwrap()
                .to_digit(10)
                .unwrap();
        }
        
        // Look for digit words
        for (i, digit_string) in DIGIT_STRINGS.iter().enumerate() {
            let found = line.find(digit_string);
            if let Some(found) = found {
                if found < first_digit_index {
                    first_digit = i as u32 + 1;
                    first_digit_index = found;
                }
            }
        }

        // Repeat above steps using line.rfind to find last digit/digit word

        let mut last_digit_index = line
            .rfind(|c: char| c.is_ascii_digit())
            .unwrap_or(usize::MAX);
        if last_digit_index != usize::MAX {
            last_digit = line
                .chars()
                .nth(last_digit_index)
                .unwrap()
                .to_digit(10)
                .unwrap();
        }
        
        for (i, digit_string) in DIGIT_STRINGS.iter().enumerate() {
            let found = line.rfind(digit_string);
            if let Some(found) = found {
                if last_digit_index == usize::MAX || found > last_digit_index {
                    last_digit = i as u32 + 1;
                    last_digit_index = found;
                }
            }
        }

        // Add to total
        total += (first_digit * 10) + last_digit;
    }

    println!("Result {total}");
}
